### DISCUSSION TITLE
scv.tl.recover_dynamics in stochastic mode

### DISCUSSION THREAD

**Discussion by kxxxjo** [theislab/scvelo#886] [Q&A] [answered]
Hi, @WeilerP 

Thanks for developing useful tool!

I have a question for "scv.tl.recover_dynamics".

I tried to analyze velocity analysis using two modes 1) stochastic and 2) dynamical.

but, the results from stochastic and dynamical are different. The stochastic results are similar to our expectation.

When I run with stochastic mode, I use scv.tl.recover_dynamics to get additional insights such as latent time.

Is it ok to use scv.tl.recover_dynamics in stochastic mode?

Thanks!

Best,
Kyungjo

**Comment by WeilerP** [comment, marked as answer]
@kxxxjo, sorry, I'm not sure I follow.

> but, the results from stochastic and dynamical are different. The stochastic results are similar to our expectation.

Given that the stochastic and dynamical model give you different results suggests to me that there might be an issue with the data. You should double-check that the phase portraits have the expected/required form. See _e.g._ #216 or #462. Also, you should not overinterpret/solely rely on projected velocities onto a low dimensional representation of the data (I assume this is what you mean by 
"results are similar to our expectation"). Instead, it is advised to quantify your results using [CellRank](https://cellrank.readthedocs.io/en/stable/), for example.

> Is it ok to use `scv.tl.recover_dynamics` in stochastic mode?

There is no way to "use `scv.tl.recover_dynamics` in stochastic mode". The function `recover_dynamics` uses the dynamical model to infer the full set of model parameters. This is not possible with the stochastic model.

**Comment by kxxxjo** [comment]
Thanks for quick reply!

To understand my situation, I add the results of both modes.

![image](https://user-images.githubusercontent.com/88973914/173045446-e941e397-86e8-46cb-9403-59bbf5532d49.png)

I already used cellrank, the results of cellrank also different.

I use scv.tl.recover_dynamics in stochastic mode using following codes:

```
# Preprocess the Data
scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)
scv.pp.moments(adata, n_pcs=30, n_neighbors=30)

# Estimate RNA velocity
scv.tl.velocity(adata, mode="stochastic")
scv.tl.velocity_graph(adata)

# Project the velocities
scv.pl.velocity_embedding_stream(adata, basis='X_umap', color="assigned_celltype", legend_fontsize=5)

scv.tl.recover_dynamics(adata, n_jobs=8)
scv.tl.latent_time(adata)
scv.pl.scatter(adata, color='latent_time', color_map='gnuplot', size=80)
```
I first run scv.tl.velocity(adata, mode="stochastic") and then scv.tl.recover_dynamics(adata, n_jobs=8). so I think it is possible with the stochastic mode. 

Thanks!

Best,
Kyungjo

**Reply by WeilerP** [reply]
`scv.tl.velocity(adata, mode="stochastic")` calculates velocities using the stochastic model. This will also estimate feasible genes and mark them as "velocity genes". `scv.tl.recover_dynamics(adata, n_jobs=8)` will then use these genes to estimate the full set of model parameters. This is **not** using the stochastic model!
As I suggested, you should look at the phase portraits to check if RNA velocity analysis is even feasible in your case. You should also ensure that the time scale of your system is comparable to the time scale of splicing kinetics.
FYI: It'd be good to answer within the same thread to make following the discussion easier and eventually mark the answer.

**Reply by kxxxjo** [reply]
thanks for nice response!

so, you mean running `scv.tl.velocity(adata, mode="stochastic")` and then `scv.tl.recover_dynamics(adata)` is not correct way?

As you suggested, I tried to look at the phase portraits but I don't know how to interpret the results although I refer to tutorial page. what genes should I input when generating phase portraits? could you let me know how to interpret them, please?

<img width="995" alt="image" src="https://user-images.githubusercontent.com/88973914/173190458-c2c6e4dc-fe8f-43c1-95d2-e43580ade4c1.png">

When I generate the phase portrait for Ovol3, the error occurred : `TypeError: cannot perform reduce with flexible type`

and how should I interpret the results?

I have another question. I want to generate the phase portraits for genes of interest but when I input the genes of interest, there were no velocity plot. I think it's because genes of interest are not included in velocity genes, so I can't see about genes of interest?
<img width="745" alt="image" src="https://user-images.githubusercontent.com/88973914/173190683-7e98555a-dcce-40af-8b64-612e15b39204.png">

Thanks!

Best,
KJ

**Reply by WeilerP** [reply]
@kxxxjo, as suspected and outline in detail in #216 or #462, your phase portraits do not exhibit the desired form and, thus, RNA velocity cannot be inferred. Please have a look at the mentioned issues and the paper for more details.

> so, you mean running scv.tl.velocity(adata, mode="stochastic") and then scv.tl.recover_dynamics(adata) is not correct way?

There is nothing wrong with it. I am simply saying that this will infer the velocities using the dynamical model.

> As you suggested, I tried to look at the phase portraits but I don't know how to interpret the results although I refer to tutorial page. what genes should I input when generating phase portraits?

You can choose genes that you know are important for the biological process you are studying as well as genes that scored a high likelihood.

**Reply by kxxxjo** [reply]
I'm sorry for late response and thanks for meaningful reply!

I would summarize the contents of ours.

1. The results of using stochastic mode seem to be our expectation, but it is not consistent with dynamical model so you suggest that I should double-check them using phase portrait because you think there was an issue with my data, right?

2. Run `scv.tl.velocity(adata, mode="stochastic")` to get velocity genes using stochastic mode and then run `scv.tl.recover_dynamics(adata)` to estimate velocity genes using full set of model parameters. This flow doesn't have any problem, right?

The terminal states (from cellrank) are different from stochastic and dynamical mode. Our expectation is the thing of stochastic mode. I wanted to generate pseudotime heatmap for terminal states of stochastic mode, so I ran `scv.tl.recover_dynamics(adata)` after running `scv.tl.velocity(adata, mode="stochastic")`.

I will double-check the phase portraits referring to #216 and #462 and update them.

Thanks!

Best,
KJ
